Enhancements to terraform-aws-ec2

JonRoma opened the pull request:
* Retire `tier` argument.
* Eliminate limitation with EC2 in private IP space.

JonRoma committed:
Retire `tier` argument

Retire deprecated `tier` argument, and remove ill-advised subnet_type
default value of "public".

JonRoma commented:
@kwessel and I came up with this PR this afternoon to address a situation involving Scrum Team D in order to allow placement of an EC2 instance on a campus-facing subnet *without* having a public IP address, instead accessing via campus private (RFC1918) address space.

@tzturner, does the current PR provide a more complete and less manual way of addressing the use case that you submitted in #21, or do they address separate scenarios?

kwessel commented on README.md:
```diff
@@ -136,15 +136,14 @@ to the EC2 instance.
-
 * `private_ip` - (Optional) Private IP address to associate with the instance in a VPC.
 
+* `use_private_ip - (Optional) Route53 record should use private rather than public IP address.
```
@JonRoma , this setting should be an attribute of the alias block. If I'm reading this right, you have it documented as a top-level setting in the inputs block, not a parameter tha goes inside an alias block.

JonRoma replied:
@kwessel, I think you are confusing the top-level `use_private_ip` argument that we added with the `private_zone` attribute that's part of the `alias` block.

I suppose one could make an argument that the choice between the private *vs.* public IP could be on a per-alias basis, but I frankly can't think of a use case that compels using multiple Route53 aliases for a single EC2 instance, and such a use case would *only* be functional for a service deployed in the "campus" subnets.

Going a step farther, I'm not sure if @ddriddle and I had a specific use case in mind when we designed the module to support multiple Route53 aliases, or whether this feature was added to the module simply because we could.

Since I'm off on medical leave, I'm not going to merge this now, and hopefully we will hear from @edthedev's team about the question posed earlier in this PR's conversation.

kwessel replied:
Nope, I wasn't talking about the private_ip setting. I was talking about the new use_private_ip that we added yesterday. I still think this is confusing. The hostname and domain as well as the Boolean private_zone setting all go in an alias block, yet what address to assign to that alias (the use_private_ip setting) goes outside that block and thus applies to all aliases. I can't think of a strong use case when an EC2 instance would need multiple aliases. It would be possible, given that the public and private IPs are assigned to different network interfaces on teh EC2 box, to have different things responding publically vs. privately. While I can't think of why someone wuld want to do that with an EC2 box, it's within the realm of possibilities, and in that case, you'd want two aliases: one being a hostname that resolves to the public IP and another hostname that resolves to the private IP. To do that, the use_private_ip, like the private_zone setting, woudl need to be inside the alias block and apply specifically to that alias.

If you don't wish to support this, then I'd encourage you to get rid of the alias block, assuming that's not a requirement of a downstream AWS module, and put hostname, domain, and private_zone as direct settings of the inputs block.

And holy cow, I can't believe you're up and alert and reading emails, Jon! Way to go, my friend!

JonRoma replied:
@kwessel, my response was relating to what you said:

> If I'm reading this right, you have it documented as a top-level setting in the inputs block, not a parameter that goes inside an alias block.

That's because the `use_private_ip` *is* defined in the top-level of the inputs block, unlike the other setting.

It would not be hard to move the new parameter into the `alias` block, and I am open to that idea. But I wanted to hear @ddriddle's views as well as those of @tzturner, who submitted the PR that added the new `private_ip` relatively recently.

The pertinent points to discuss are:

**When we originally coded this module, did we actually identify a use case that justifies having an `alias` block at all, or did we just do this because we could?**

I don't like supporting code that makes things needlessly complex if we gain nothing from that complexity. If we were starting anew, I would not implement multiple domain/hostname pairs as we did without good reason.

However, given the status quo, I'm not entirely eager to rip out working code implementing the `alias` block by placing the domain/hostname as top-level input variables. If we *really* wanted to gold plate it, the alias block could be a map containing domains (zones) as keys with a list of hostnames as values. :-)

**With the new capability we added yesterday, does SecDev still have a use case for the `private_ip` that was added, or does our new feature resolve that?**

I do not know what their needs are.

**We should think about the ambiguity in naming the arguments `private_ip`, `use_private_ip`, and the `private_zone` attribute of the `alias` block.**

This is potentially confusing to the wider audience we hope to reach, so now is a good time to consider naming or at least improving the documentation.

The reason I want more opinions on this is that I am on the fence.

JonRoma replied:
@kwessel wrote:

> And holy cow, I can't believe you're up and alert and reading emails, Jon! Way to go, my friend!

That's because my procedure doesn't take place till Friday. :-)

ddriddle replied:
@JonRoma I hope you procedure goes well and you feel well soon!

I agree with @kwessel that the new argument `use_private_ip` belongs in the alias block. At least from a UI perspective. 

> Going a step farther, I'm not sure if @ddriddle and I had a specific use case in mind when we designed the module to support multiple Route53 aliases, or whether this feature was added to the module simply because we could.

I had little to do with the development of this particular module. I'm not sure why you added support for multiple alias blocks but I suspect it is because we did this elsewhere and you copied code from another module, or maybe it is something that Maiko requested at the time. Whatever the initial reason, I think it is a good feature and worth keeping if the work to maintain it is not too onerous. In general, I prefer generic solutions to overly specific solutions unless there is a good reason such as maintenance cost or performance.

JonRoma replied:
@kwessel and @ddriddle, you've convinced me to get off the fence and make `use_private_ip` an attribute of `alias`. After thinking about it for a while, this makes sense because the choice of public *vs.* private IP address is tied to an alias record.

However, this doesn't necessarily produce as winning combination as we think. See the following snippet:

```
# Route53 records created for the EC2 instance.
  alias = [
    {
      hostname       = "identity-crisis"
      domain         = "test.iamu.techservices.illinois.edu"
      use_private_ip = true
    },
    {
      hostname       = "crisis-of-identity"
      domain         = "test.iamu.techservices.illinois.edu"
      use_private_ip = false
    }
  ]

  subnet_type                 = "campus"
  associate_public_ip_address = true
```
From outside campus IP space (i.e., when I am at home and **not** on the VPN), I can't SSH to the box using **either** of the aliases:

1. The private IP is inaccessible because campus won't route off-campus traffic into RFC1918 space.
1. The public IP is inaccessible because a campus-or private-facing subnet doesn't accept inbound traffic from outside the VPC.

Using a private IP for a resource in the "public" `subnet_type` won't work either for reason 1. Likewise, using a public IP for a resource in the "private" `subnet_type` also fails for reason 2.

If I read the situation correctly, supporting multiple aliases is worthless unless **all** aliases have the same value for the `use_private_ip` boolean **and** are in the "campus" `subnet_type` **and** the would-be SSH user is on campus (which includes being on the VPN).

The support for multiple aliases now strikes me as creating confusion without providing meaningful value. I am beginning to think that the "flexibility" of supporting multiple aliases is a mirage because of network routing.

Rather than ditching the `alias` block, I wonder if it should instead become a simple map rather than a list of maps. This would continue to group the related arguments together without building false expectations. Here is what I suggest: 

```
# Route53 records created for the EC2 instance.
  alias = {
    hostname       = "identity-crisis"
    domain         = "test.iamu.techservices.illinois.edu"
    use_private_ip = true
  }
```

Thoughts?

JonRoma replied:
I'm also interested in feedback on [the questions posed in this comment](https://github.com/techservicesillinois/terraform-aws-ec2/pull/23#discussion_r1019734964). I'm looking to implement this as soon as we have an adequate discussion since we can nearly immediately deploy code using a public IP address in a campus-facing subnet.

kwessel replied:
> I'm also interested in feedback on [the questions posed in this comment](https://github.com/techservicesillinois/terraform-aws-ec2/pull/23#discussion_r1019734964). I'm looking to implement this as soon as we have an adequate discussion since we can nearly immediately deploy code using a public IP address in a campus-facing subnet.

kwessel replied:
> 2. The public IP is inaccessible because a campus-or private-facing subnet doesn't accept inbound traffic from outside the VPC.

Is that true just because of the routing in our campus-facing subnet? Or is that inherent to any type of campus-facing subnet?

I can understand this argument, but I can also see a campus-facing subnet that's designed to also accept inbound public traffic. If that's doable, then assigning both private and public IP addresses to hostnames could be useful. Perhaps this is way too much of a niche case and I'm overthinking this. However, the code's there and working, and the only issue is the current routing of the subnet in which it's deployed. In a different situation, this could work. My vote woudl be to leave it as-is, but I'm willing to go either way on this as you all feel fit.

JonRoma replied:
@kwessel wrote:

> > 2. The public IP is inaccessible because a campus-or private-facing subnet doesn't accept inbound traffic from outside the VPC.
> 
> Is that true just because of the routing in our campus-facing subnet? Or is that inherent to any type of campus-facing subnet?

Far be it from me to attempt to explain networking with any degree of competence, but the behavior is described by David Z's [Amazon Web Services, VPC Guide for Illinois](https://answers.uillinois.edu/illinois/page.php?id=71015#subnet-types).

In the table below reproduced from that document (omitting IPv6), here is a table that may or may not pass an accessibility test.

  | Public-facing subnet | Private-facing subnet | Campus-facing subnet
-- | -- | -- | --
To / from private IPv4 in same VPC | yes | yes | yes
To / from private IPv4 in other VPC | yes (via Transit Gateway or VPC peering) | yes (via Transit Gateway or VPC peering) | yes (via Transit Gateway or VPC peering)
Outbound IPv4 to Internet (or any publicly reachable server) | yes (requires public IP or Elastic IP) | yes (via NAT Gateway in your VPC) | yes (via NAT Gateway in your VPC or shared NAT Gateway in Core Services VPC)
Inbound IPv4 from Internet (including campus clients) to AWS public IP | yes (requires public IP or Elastic IP) | no | no
Outbound to campus host (not publicly reachable) from AWS private IPv4 | no | no | yes
Inbound from campus host to AWS private IPv4 | no | no | yes

Summing up the answer to your question, inbound IPv4 traffic from the Internet (including campus clients) is **not** routed to campus-facing and private-facing subnets. However, such inbound traffic *is* routed to a public-facing subnet, provided the destination address has a public IP or elastic IP.

David Z. went on to say that 
> Campus-facing subnets support bi-directional communication without NAT to campus networks at all three universities in the U of I System (some exceptions apply, see the Detailed Example).  They are otherwise similar to private-facing subnets, but currently support only IPv4 (not IPv6).

I interpret this as a statement that neither campus- nor private-facing subnets route traffic from the public Internet – and that, accordingly, you can only access an EC2 box placed in a campus-facing subnet from within the VPC or from a campus source address.

kwessel replied:
Right, we're all already clear on how it works. That's how it's defined right now. That's not to say that additional public routing rules couldn't be added in the future. Clearly, nothing in the campus VPC would work with a public IP. Outbound traffic is NAT'd if it's going over the public internet, and inbound traffic won't work. So, to follow the argument you're making, @JonRoma , of saying that the current state of this code could lead to confusion would also argue against things like assign_public_ip in the campus facing subnet. I think the code, as it is right now, allows for maximum flexability, and it's up to the deployer to use smarts as to the routing rules of the VPC in which they're deploying. Should additional routing rules be added to campus-facing subnets in the future, it leaves room more more flexability with the code that's currently there. I don't think the current state of things hurts anything.

JonRoma committed:
Allow creating EC2 instances in private IP space

Use of the existing `private_zone` attribute inside the `alias`
block allowed creating a Route53 record in a private (per-VPC) zone.
However, it was previously not possible to create an EC2 instance in
campus private (RFC1918) IP space with a *public* Route53 record.

This change allows the creation of EC2 instances that can be accessed
via SSH from campus via the private IP address, provided the instance's
security group configuration permits.

This commit adds the `use_private_ip` argument to the `alias` block.

Note that this module supports creating arbitrarily many alias
records using either public or private IP addresses. Depending on
the subnet placement of the EC2 instance, not every possible
configuration will actually be routeable.

ddriddle commented on README.md:
```diff
@@ -136,15 +136,14 @@ to the EC2 instance.
 
 * `template_file` - (Optional) User data template file.
 
-* `tier` - (Optional) A subnet tier tag (e.g., public, private,
```
Where is the documentation for the `subnet_type`?

JonRoma replied:
It was missing.

ddriddle commented:
> @kwessel and I came up with this PR this afternoon to address a situation involving Scrum Team D in order to allow placement of an EC2 instance on a campus-facing subnet _without_ having a public IP address, instead accessing via campus private (RFC1918) address space.
> 
> @tzturner, does the current PR provide a more complete and less manual way of addressing the use case that you submitted in #21, or do they address separate scenarios?

This PR does not address our use case. We have no need for a route53 record at this time. Our use case is that we need to manually specify a private static IP for our EC2 box. We need this because the static private IP is whitelisted to access secured services on campus.

kwessel commented:
It's possible that David's team's need and Scrum Team D's need might both work together. Scrum Team D needs a hostname they can ssh to. They also have traffic going to services on campus that might be firewalled more restrictively than our whole VPC subnet. I maintain that those service should be able to open things up to our whole VPC subnet. But if not, they could reserve a static private IP the way David's group is and plug it into the EC2 definition while also letting the EC2 definition create the hostname. @ddriddle , is this overkill? Is it better, if you're reserving a static private IP, to just manually assign the hostname in Route53 when you reserve it?

ddriddle reviewed (dismissed):
Please document `subnet_type`. 

@kwessel and @JonRoma I don't understand the need for `use_private_ip`. Why would you want to assign a private IP to a public DNS record? Is that even valid? Even if possible, it seems like an abusive hack. It seems to me if you are using a private IP in campus space you would use a private zone, or create an A record in the campus DNS system. You might want to ask David Z. for a recommendation, but I don't think this solution is ideal.

If the only reason you are doing this is for SSH, then I recommend that each developer configure their `/etc/hosts` file or `~/.ssh/config` to create an alias. For static IP(s) this is not so bad, but for dynamic records this would be inconvenient. If you do not want to deal with static IP(s), you probably could configure the developer machines to resolve AWS private records:

https://aws.amazon.com/premiumsupport/knowledge-center/client-vpn-resolve-resource-records/

ddriddle commented:
@JonRoma and @kwessel instead of putting a private IP in a public DNS record, you could write a script that looks up a private hosted zone record using the AWS API then write that value into `/etc/hosts` or `~/.ssh/config` or wherever the developers want it to go. If you create such a script please share it!

Of course, I'm sure David Z. can recommend a more elegant solution that uses campus IPAM and private zones.

kwessel commented:
@ddriddle , the fact that you had to ask this question tells me we didn't make our intention clear in the first place. The idea is this: we want to assign a hostname to the campus facing IP. Now that we have the campus-facing subnets, those 10.x.x.x addresses can be used from campus to access services. In the case of the primary driver here, the Scrum Team D EC2 work environment, we want a box that they can ssh into thrugh the VPN. Because that privatenet space is part of campus IP space, it's accessible from any on-campus client including VPN clients. It also means the system doesn't need a public IP. It doesn't need a list of CIDR blocks for off-campus access to be established, and people accessing the system don't have to use the tunnel all VPN profile. And it means the system can be treated as fuly on-campus, even with its outbound traffic to other on-campus services. Does that clarify things?

kwessel commented:
It's also wworth pointing out that we have plenty of private IPs in campus DNS. Any of our 172.16.X.X systems have DNS entries. Those may only be resolvable from campus, mind you, but the concept stil applies. If there's anything inherently wrong about this, I'd love for david Z to let us know that. I don't see this as an issue.

Your external script is any interesting idea, but it feels like an extra unnecessary hook to get things working.

ddriddle commented:
@kwessel is this for production workloads or just development? I think the simplest solution is to hardcode a static IP in the Terragrunt configuration, then create a record in campus IPAM particularly for production workloads. For development you could create a script or just hardcode the IP. That is the route we went. Again, I don't like the idea of creating a public DNS record that contains a private IP. That is what you are trying to do, correct?

For our use case, it is a dev box we SSH into using the campus VPN. We IP restricted the EC2 box (using security groups) so it is only accessible from two factored authenticated VPN users. We did not want to open it up to all of campus. We just did not see a need for a Route53 or campus IPAM DNS record since the IP is private & static, nor should it be accessed by anyone outside our team.

kwessel commented:
Yes, that is what we're trying to do: put a 10.X.X.X campus-facing IP in the Route53 public zone. I agree that there are several ways to accomplish this, but I still fail to see the problem with putting a private IP in public DNS. Can you explain your concern with that? It's by far the easiest solution.

If you really don't want it to be public, we could put the A record in a public zone that IPAM delegates to Route53 but only for on-ampus queries, I suppose. But I still don't understand the issue.

JonRoma committed:
Document missing subnet_type argument

JonRoma commented:
Regarding private IP addresses, manually assigning an address rather than letting AWS assign it automatically does not seem attractive.

Regarding name resolution, there are numerous ways to provide name resolution for addresses in campus private space. With all due respect, none of the alternatives proposed aren't particularly attractive. For example, dealing with the campus IPAM web interface is unpalatable, both for vision-impaired people who depend on a screen reader, as well as people without vision impairment. It is an annoying, extra manual step.

Unless Host Manager disfavors the approach we took – and maybe we should ask – the solution @kwessel and I found seems to be the most straightforward.

kwessel commented:
Yeah, I don't think vision or no vision has anything to do with it. The genral consensus is that the IPAM interface sucks. Still need to play with the Terraform interface to Infoblox.

JonRoma commented:
> The genral consensus is that the IPAM interface sucks. Still need to play with the Terraform interface to Infoblox.

Yeah, I wonder if David Z. has explored that Terraform provider. It would be worth a look.

JonRoma commented:
> I can understand this argument, but I can also see a campus-facing subnet that's designed to also accept inbound public traffic. If that's doable, then assigning both private and public IP addresses to hostnames could be useful. Perhaps this is way too much of a niche case and I'm overthinking this. However, the code's there and working, and the only issue is the current routing of the subnet in which it's deployed. In a different situation, this could work. My vote woudl be to leave it as-is, but I'm willing to go either way on this as you all feel fit.

I believe my previous comment quoting David Z's documentation suggests that our campus-facing subnets unequivocably do **not** accept inbound public traffic. Hence, it is my belief that having both public and private IP addresses on a campus-facing host is not a meaningful scenario.

Regarding the Terraform data structures, the fact that working code exists is an argument in favor of leaving it alone. The flip side is that the current data structure implies more freedom to support edge cases than actually be successfully deployed. That in turn suggests that we either need to document the nuances more carefully (which could get hairy), or simplify the module to not imply more flexibility than it can actually deliver.

kwessel commented:
> > I can understand this argument, but I can also see a campus-facing subnet that's designed to also accept inbound public traffic. If that's doable, then assigning both private and public IP addresses to hostnames could be useful. Perhaps this is way too much of a niche case and I'm overthinking this. However, the code's there and working, and the only issue is the current routing of the subnet in which it's deployed. In a different situation, this could work. My vote woudl be to leave it as-is, but I'm willing to go either way on this as you all feel fit.
> 
> I believe my previous comment quoting David Z's documentation suggests that our campus-facing subnets unequivocably do **not** accept inbound public traffic. Hence, it is my belief that having both public and private IP addresses on a campus-facing host is not a meaningful scenario.
> 
> Regarding the Terraform data structures, the fact that working code exists is an argument in favor of leaving it alone. The flip side is that the current data structure implies more freedom to support edge cases than actually be successfully deployed. That in turn suggests that we either need to document the nuances more carefully (which could get hairy), or simplify the module to not imply more flexibility than it can actually deliver.

So, does that mean that you don't want someone to be able to assign a public IP to a service running in the private or campus-facing subnet? Because, as I mentioned in my previous comment, that's also doable. At some point, it comes down to simply this: the deployer needs to know what their subnet type is capible of and should try to deploy something that the subnet type can't support. As for documenting it, that's already done: you quoted from David Z's docs, and that's all that's needed. If you're concerned, simply include a note regarding anything subnet or IP assignment related in teh readme pointing people to his documentation. I'd rather see that than you li miting functionality to prevent the proverbial shot in the foot.

JonRoma commented:
> That's not to say that additional public routing rules couldn't be added in the future. Clearly, nothing in the campus VPC would work with a public IP. Outbound traffic is NAT'd if it's going over the public internet, and inbound traffic won't work. 

@kwessel, as things stand today, the three subnet groups (campus, private, and public) and enabling NAT are all features inherently supported by David Z's module. All I really do is invoke his code with the proper input values (and with the addition of tags to allow us to look up rather than hardcode subnet IDs).

You mentioned the possible addition of new public routing rules in the future. Just so we're not misunderstanding one another, I wish to clarify that I do not intend to support networking features that are not part of David Z's module. This is too much akin to heading into deep water without a life raft.

If, on the other hand, David Z. chooses to support new features with new versions of his Terraform module that interest IAMU, I am all ears – though you are likely aware that reorganizing or replacing IAMU's VPC will be time-intensive and disruptive. So we'll have to weight the cost/benefit ratio if that ever came up.

JonRoma commented:
> So, to follow the argument you're making, @JonRoma , of saying that the current state of this code could lead to confusion would also argue against things like assign_public_ip in the campus facing subnet. I think the code, as it is right now, allows for maximum flexability, and it's up to the deployer to use smarts as to the routing rules of the VPC in which they're deploying. Should additional routing rules be added to campus-facing subnets in the future, it leaves room more more flexability with the code that's currently there. I don't think the current state of things hurts anything.

I'll admit that, very early on, the circumstances wherein one needs `assign_public_ip` seemed non-intuitive to me (and I think also to @ddriddle). On the other hand, that was *long* before David Z's `aws-enterprise-vpc` module was revamped. The fix we made last week to allow an EC2 box to reside in campus-facing private IP space is definitely a win that fits a real-life use case for Scrum Team D and possibly others.  I'm happy that we explored that, and that doesn't require more than one alias record.

To answer your question: the **sole** basis of my distaste for supporting multiple aliases is simply that one can create multi-alias configurations that are dysfunctional because the routing isn't supported by the underlying network configuration. I think we're shooting ourselves in the foot by supporting something that involves a lot of ifs, ands, or buts. This becomes especially problematic and confusing if we spread this code across campus to people who are even less network-knowledgeable than I! FYI, the `terraform-aws-ec2` module is *public*.

I'm not looking for another project, but I wonder if it's worth seeing which combinations work and which don't. Here are the variables:
* three mutually exclusive subnet types.
* `assign_public_ip` boolean.
* `use_private_ip` boolean.
* NAT gateway defined (boolean).
* On campus or not (boolean).

If we were starting anew, I wouldn't go down the road of multi-alias configurations. However, obviously, there is zero  development effort needed to support multi-alias configurations since working code already exists. So I'm not going to dig in my heels on this; I just like to keep sight on what people actually *need* rather than do things simply because we can – especially if it creates support headaches.

@edthedev, I wouldn't mind a look from a neutral third party like you, especially since I know you are looking ahead at a larger audience than simply Tech Services. FYI, this module is public.

The discussion hinges on a data structure described [earlier in this discussion](https://github.com/techservicesillinois/terraform-aws-ec2/pull/23#discussion_r1022385530). That comment shows both data structures. The present object is a Terraform list of maps that supports multiple Route53 aliases. The alternative is a Terraform map supporting a single alias.

kwessel commented:
> You mentioned the possible addition of new public routing rules in the future. Just so we're not misunderstanding one another, I wish to clarify that I do not intend to support networking features that are not part of David Z's module. This is too much akin to heading into deep water without a life raft.

My turn to clarify: I never intended to imply that _you_ or any of us would add networking rules. It was more a general statement of the networking modules being further enhanced, and yes, agreed that David Z. would be making those enhancements if the need arose, not us. So, we're in agreement on this point.

JonRoma commented:
> > You mentioned the possible addition of new public routing rules in the future. Just so we're not misunderstanding one another, I wish to clarify that I do not intend to support networking features that are not part of David Z's module. This is too much akin to heading into deep water without a life raft.
> 
> My turn to clarify: I never intended to imply that _you_ or any of us would add networking rules. It was more a general statement of the networking modules being further enhanced, and yes, agreed that David Z. would be making those enhancements if the need arose, not us. So, we're in agreement on this point.

All right; thanks. That's what I had *hoped* you meant.
